script/tool/lib/src/federation_safety_check_command.dart

```diff
@@ -0,0 +1,187 @@
+          relativeComponents.first == 'pubspec.yaml' &&
+          await _packageWillBePublished(path)) {
+        _modifiedAndPublishedPlatformInterfacePackages.add(packageName);
+      }
+    }
+  }
+
+  @override
+  Future<PackageResult> runForPackage(RepositoryPackage package) async {
+    if (!isFlutterPlugin(package)) {
+      return PackageResult.skip('Not a plugin.');
+    }
+
+    if (package.directory.parent.basename == 'packages') {
+      return PackageResult.skip('Not a federated plugin.');
+    }
+
+    if (package.directory.basename.endsWith('_platform_interface')) {
+      // As the leaf nodes in the graph, a published package interface change is
+      // assumed to be correct, and other changes are validated against that.
+      return PackageResult.skip(
+          'Platform interface changes are not validated.');
+    }
+
+    // Uses basename to match _changedPackageFiles.
+    final String basePackageName = package.directory.parent.basename;
+    final String platformInterfacePackageName =
+        '$basePackageName$_platformInterfaceSuffix';
+    final List<String> changedPlatformInterfaceFiles =
+        _changedDartFiles[platformInterfacePackageName] ?? <String>[];
+
+    if (!_modifiedAndPublishedPlatformInterfacePackages
+        .contains(platformInterfacePackageName)) {
+      print('No published changes for $platformInterfacePackageName.');
+      return PackageResult.success();
+    }
+
+    if (!changedPlatformInterfaceFiles
+        .any((String path) => path.startsWith('lib/'))) {
+      print('No public code changes for $platformInterfacePackageName.');
+      return PackageResult.success();
+    }
+
+    // If the change would be flagged, but it appears to be a mass change
+    // rather than a plugin-specific change, allow it with a warning.
+    //
+    // This is a tradeoff between safety and convenience; forcing mass changes
+    // to be split apart is not ideal, and the assumption is that reviewers are
+    // unlikely to accidentally approve a PR that is supposed to be changing a
+    // single plugin, but touches other plugins (vs accidentally approving a
+    // PR that changes multiple parts of a single plugin, which is a relatively
+    // easy mistake to make).
+    //
+    // 3 is chosen to minimize the chances of accidentally letting something
+    // through (vs 2, which could be a single-plugin change with one stray
+    // change to another file accidentally included), while not setting too
+    // high a bar for detecting mass changes. This can be tuned if there are
+    // issues with false positives or false negatives.
+    const int massChangePluginThreshold = 3;
+    if (_changedPlugins.length >= massChangePluginThreshold) {
+      logWarning('Ignoring potentially dangerous change, as this appears '
+          'to be a mass change.');
+      return PackageResult.success();
+    }
+
+    printError('Dart changes are not allowed to other packages in foo in the '
```

Comment:
foo?

Reply:
Turns out that when writing the error message's test first, and copy/pasting it to the implementation rather than rewriting it, it's a good idea to remember to swap the variables in.